### ISSUE TITLE
Proposal; Create a small web service to pick a random pif.json and load it at boot time.

### ISSUE THREAD

**Issue by betterphp** [chiteroman/PlayIntegrityFix#179]
Gonna try and move the discussion about this away from the forum thread as it's pretty hard to follow anything on there atm!

I propose we create a small API to return a randomly selected pif.json from a list of known working ones. The list would be stored in a database, not the github repo.

Motivation for doing this;
* Avoids everyone having to spend time hunting for the same things.
* We could take steps to block Google from accessing the API, and even if they did the full list would still be secret. Obviously there would be no way to fully prevent them accessing the API but if we could try and make it hard enough for them not to bother.
* A more even distribution of fingerprint usage might mean they stop getting banned all together.

The API code itself would be open source and I could create and host it. We could use github issues to allow people to report fingerprints that stop working but I'd need a handful or working ones to populate it to start with.

We could explore other options for the selection too; return the ones that have been used the least number of times in the last day or something like that maybe - although I'd suggest just random to start with for simplicity.

## Request example
```
GET https://whatever.com/pif.json
```

### Success (HTTP 200)
This one is the most obvious, a suitable value was found.
```json
{
    "status": "success",
    "pif_json": {
        "PRODUCT": "b1-780_ww_gen1",
        "DEVICE": "acer_barricadewifi",
        "MANUFACTURER": "Acer",
        "BRAND": "acer",
        "MODEL": "B1-780",
        "FINGERPRINT": "acer/b1-780_ww_gen1/acer_barricadewifi:6.0/MRA58K/1481784106:user/release-keys",
        "FIRST_API_LEVEL": "21"
    }
}
```

### Error rate limited (HTTP 429)
If an IP address requests more than X values per hour, probably with an exponential backoff time.
```json
{
    "status": "error",
    "message": "Rate limit exceeded"
}
```

### Error you don't look like a mobile phone (HTTP 403)
If something suggests the requests is not coming from a phone. There's no totally foolproof way to do this one but stuff like the owner of the IP address could be used.
```json
{
    "status": "error",
    "message": "Access denied"
}
```

Thoughts?

**Comment by Iridule** [comment]
> Thoughts?
I don't think that'll work ppl would be so frustrated having to randomly clear gps data and wallet so often...

**Comment by non-bin** [comment]
I think this would be great, I wonder does anyone know if the module could automatically clear wallet data so you don't have to do it manually?

**Comment by betterphp** [comment]
> > Thoughts?
> > I don't think that'll work ppl would be so frustrated having to randomly clear gps data and wallet so often...

I think you'd only need to clear the data if your device failed a validation and cached the result, not every time the fingerprint changed.

When NFC payments stopped working on my device after the 14.4 update all I had to do was clear the cache (not data) for google play services.

**Comment by furdiburd** [comment]
> I think this would be great, I wonder does anyone know if the module could automatically clear wallet data so you don't have to do it manually? I've wanted to contribute to this for a while but I don't really have the knowledge, so I'm happy to help develop the API, or donate to help host it

I think this whole wipe gms and wallet data is unecessary. Wallet will just do the authentication when you try to pay and the warning of not meeting security just go away the next day. Ofc google play store instantly work like its needed to (netflix can be downloaded) after installing a version with unbanned fingerprint so auto wiping data would be just an unnecessary problem

**Comment by ZerkerEOD** [comment]
I'm seriously contemplating moving to iOS after all the crap Google has been doing recently to play fuck fuck games with it's customers that willingly root their devices...

**Comment by furdiburd** [comment]
> I'm seriously contemplating moving to iOS after all the crap Google has been doing recently to play fuck fuck games with it's customers that willingly root their devices...

The problem is... Android is better one even if not good. Can you use third party apps on ios? Nope. Can you easily root it? Nope. :/

**Comment by non-bin** [comment]
> contemplating moving to iOS 

Yeah, cos apple is way better with allowing customisation

**Comment by notheowner** [comment]
As long as you are making it easy for the average Joe to use this mod, Google will keep banning prints. 
This whole "We can try to take measure against Google" just won't work. Stop automating this process more, it's already dumbed down enough where people all they do is flash a public print to get it banned in 3 days. We don't need to speedrun fingerprint bans. Let people fend for themselves. These are not solutions.

**Comment by betterphp** [comment]
I guess just remember that all of these are actually "public prints", we're just individually spending time to find the same lists of working values.

This is how i see it;

1. in half an hour someone could maybe find one or two
2. 100 people searching in privately could find the same 1 or 2
3. 100 people searching to add to the same list could find 100 - 200

With the second option the two easy to find fingerpints will be used nearly as much as the one in this repo.

With the third option each of those would only be used once more than normal on average.

**Comment by betterphp** [comment]
We'll need to hear what @chiteroman thinks on this before doing anything anyway.

**Comment by Saduff** [comment]
> A more even distribution of fingerprint usage might mean they stop getting banned all together.

Do we have any data that the reason the prints are getting banned is due to high usage? Like, is there anyone who's been using a custom print for a while that hasn't been banned? Because if those get banned as frequently as the ones included in the module, we have a different problem and this solution won't change anything.

**Comment by furdiburd** [comment]
> > A more even distribution of fingerprint usage might mean they stop getting banned all together.
> 
> Do we have any data that the reason the prints are getting banned is due to high usage? Like, is there anyone who's been using a custom print for a while that hasn't been banned? Because if those get banned as frequently as the ones included in the module, we have a different problem and this solution won't change anything.

I think google just check the fingerprint this module currently use and ban it and not check how many device use the fingerprint.

**Comment by betterphp** [comment]
> Do we have any data that the reason the prints are getting banned is due to high usage? Like, is there anyone who's been using a custom print for a while that hasn't been banned?

Only anecdotally by people bragging on the XDA thread 😆

We'll only be able to get that kind of data by experimenting with widely used modules like this. I suspect that's why we have a new fingerprint in 14.6 along with a bunch of code changes and extra props being set.

**Comment by non-bin** [comment]
This is awesome! Don't you think that having the PIFs public like this will get them all banned?

**Comment by furdiburd** [comment]
> This is awesome! Don't you think that having the PIFs public like this will get them all banned?
> 
> EDIT: Just had a look at the hosted version, is this just the script from the repo? I think it could also do with some more strict rate limiting, like 3 per day kind of strict. With the current limit, how long would it take to download the whole list?

5 per day is the limit. Checked it. Some of those alresdy banned actualy, we need a way to mark jsons as banned/untested

**Comment by furdiburd** [comment]
> > This is awesome! Don't you think that having the PIFs public like this will get them all banned?
> > EDIT: Just had a look at the hosted version, is this just the script from the repo? I think it could also do with some more strict rate limiting, like 3 per day kind of strict. With the current limit, how long would it take to download the whole list?
> 
> I didn't do a limit because it's all available from the repo

Weird, i got a "Rate limit exseeded" error after 5 refresh

**Comment by non-bin** [comment]
@herzhenr/spic-android and @herzhenr/spic-server are open source

**Comment by furdiburd** [comment]
> @herzhenr spic-android and spic-server are open source
> 
> This is a very noob question, but why can't we just use the fingerprint of say the latest pixel release or something like that? They couldn't block that or every pixel phone would stop working

Cuz new phones use hardware backed checks. Those phone fingerprints cant be used to spoof

**Comment by InfiniteFeld** [comment]
> > @herzhenr spic-android and spic-server are open source
> > 
> > This is a very noob question, but why can't we just use the fingerprint of say the latest pixel release or something like that? They couldn't block that or every pixel phone would stop working
> 
> Cuz new phones use hardware backed checks. Those phone fingerprints cant be used to spoof

So essentially there is a fixed window in time of devices that have usable fingerprints. As we continue into the future, and that fixed window gets pushed further into the past, and Google continues to ban/obsolete devices in that window, there will eventually be no more usable fingerprints and rooting will be dead just like jailbreaking iOS? Does anyone have a SWAG as to how long we have left? And what's everyone's plan for that locked down dystopian future??

**Comment by furdiburd** [comment]
GOOD NEWS! There is a fork that do exactly this!!!

https://github.com/daboynb/PlayIntegrityNEXT/tree/Release

**Comment by SpeedyCraftah** [comment]
> I'm 99% sure that Google isn't spying on this repo to block those fingerprints, unless Google is so dumb that they can't find out a high peak of integrity api calling count from a random old device model.

An even distribution of keys with a database that isn't very easily accessible could slow Google down a bit and go under the radar.

**Comment by SpeedyCraftah** [comment]
> Error you don't look like a mobile phone (HTTP 403)

We could even heavily obfuscate the API so that poor and naïve attempts at accessing the API outside of the integrity fix app will be easily detected (Google heavily makes use of this for their APIs so it's like using it against them in a sense) and a fake fingerprint could even be returned or one that's already been revoked, just to throw people off.

It's not a great solution but it's possible at least for the time being.

**Comment by chiteroman** [comment]
I do that and in one day Google bans all fingerprints.

**Comment by InfiniteFeld** [comment]
> I do that and in one day Google bans all fingerprints.

What's stopping Google from banning all fingerprints right now anyway? I read that locked bootloader devices will still work just fine even if their print is banned.  So why doesn't Google just end the cat and mouse game and ban all prints?

**Comment by non-bin** [comment]
Hang on, if I just lock my bootloader it'll pass? Even with root?

**Comment by Saduff** [comment]
> > I do that and in one day Google bans all fingerprints.
> 
> What's stopping Google from banning all fingerprints right now anyway? I read that locked bootloader devices will still work just fine even if their print is banned.  So why doesn't Google just end the cat and mouse game and ban all prints? 

Yep, the actual devices we're spoofing still pass when their fingerprint is banned. So the real reason the prints stop working is not clear. The Google 'ban' doesn't apply to real devices. To me, it suggests the spoofing is not perfect. If it was, we'd be able to simply spoof an older Pixel device and there's nothing Google can do other than dropping support for devices without hardware attestation.

**Comment by Saduff** [comment]
> Hang on, if I just lock my bootloader it'll pass? Even with root?

You can't lock the bootloader while keeping root. I mean, you can, but the phone won't boot. Or in other words, it'll be an expensive brick.

**Comment by betterphp** [comment]
> > I do that and in one day Google bans all fingerprints.
> 
> What's stopping Google from banning all fingerprints right now anyway? I read that locked bootloader devices will still work just fine even if their print is banned. So why doesn't Google just end the cat and mouse game and ban all prints?

Exactly, there are thousands of ROM dumps that include the properties that these modules set and even repos  with the compiled jsons in. The idea that these fingerprints are in any way private makes no sense.


> Yep, the actual devices we're spoofing still pass when their fingerprint is banned. So the real reason the prints stop working is not clear. The Google 'ban' doesn't apply to real devices. To me, it suggests the spoofing is not perfect. If it was, we'd be able to simply spoof an older Pixel device and there's nothing Google can do other than dropping support for devices without hardware attestation.

Exactly! There must be something more to it, some kind of mismatch in the props that get sent perhaps - but then it's weird that they ever pass in the first place, like why isn't that just part of the checks?
